### ISSUE TITLE
structopt as main args through rust-cli/paw

### ISSUE THREAD

**Issue by yoshuawuyts** [TeXitoi/structopt#187]
A few weeks ago the CLI WG released [paw](https://github.com/rust-cli/paw), a crate to try and make CLI parsing more of a first-class citizen in Rust ([post](https://blog.yoshuawuyts.com/paw/)). This allows arguments to be passed into `fn main` as long as a trait is implemented. Obviously we also added [structopt support](https://github.com/rust-cli/paw/tree/master/paw-structopt).

So far it's been working quite well, but we think we could make this even better if we could add the trait to structopt directly:

__Proposed__
```rust
#[derive(structopt::StructOpt)]
struct Args {}

#[paw::main]
fn main(args: Args) {
    println!("{}", args);
}
```

__Current__
```rust
#[derive(paw_structopt::StructOpt, structopt::StructOpt)]
struct Args {}

#[paw::main]
fn main(args: Args) {
    println!("{}", args);
}
```
## Implementation
The patch to `structopt` would be about 6 lines inside the structopt macro:
```rust
impl paw::ParseArgs for #name {
    type Error = std::io::Error;
    fn parse_args() -> Result<Self, Self::Error> {
        Ok(<#name as structopt::StructOpt>::from_args())
    }
}
```

Would you be open to a patch that allows structopt to be used directly from main?

---

## Full TCP Example
```rust
#[derive(structopt::StructOpt)]
struct Args {
    #[structopt(flatten)]
    port: clap_port_flag::Address,
    #[structopt(flatten)]
    port: clap_port_flag::Port,
}

#[paw::main]
fn main(args: Args) -> Result<(), std::io::Error> {
    let listener = TcpListener::bind((&*args.address, args.port))?;
    println!("listening on {}", listener.local_addr()?);
    for stream in listener.incoming() {
        stream?.write(b"hello world!")?;
    }
    Ok(())
}
```

**Comment by TeXitoi** [comment]
Can't you just `impl<T> ParseArg for T where T: StructOpt`

**Comment by yoshuawuyts** [comment]
@TeXitoi unfortunately we can't because we're not allowed to implement traits for foreign types (believe it's because of orphan rules):

![2019-05-27-114218_1920x1080](https://user-images.githubusercontent.com/2467194/58411380-77718200-8074-11e9-8337-2f8425c1cdd7.png)

Maybe I'm missing something tho?

**Comment by TeXitoi** [comment]
As paw-raw is very small, I'm OK adding the generic impl in the structopt crate (and not in the generated code as proposed above).

**Comment by TeXitoi** [comment]
@yoshuawuyts I let you open the PR? I don't know when I'll find the time to do that, but if you do it, it'll be faster.

**Comment by yoshuawuyts** [comment]
@TeXitoi yeah for sure; @gameldar has been working on a PR that sounds quite promising!

**Comment by gameldar** [comment]
I've just created PR #192 to address this. After trying a few different methods I had to go back to doing this in the structopt macro as the error @yoshuawuyts was seeing shows up because of the way ParseArgs is invoked. The only way I could get this working was to do it in paw which introduced a dependency on structopt for paw which given the intention to move paw to std at some point (so cannot have dependencies) is not an option.
